### DISCUSSION TITLE
2023: How are you using go-c8y-cli?

### DISCUSSION THREAD

**Discussion by reubenmiller** [reubenmiller/go-c8y-cli#298] [General]
Hi everyone,

I'm interested to hear from users how they are using go-c8y-cli in your Cumulocity IoT daily lives. I find sharing use-cases is a great way to discover new ways of using go-c8y-cli that you may not have previously considered or even known was possible.

For example, I've personally used go-c8y-cli to:

* Deploy a microservice in a CI/CD pipeline
* Troubleshooting your IoT Solution (e.g checking device logs, looking for events around a specify timestamp etc.)
* Fleet/Device management (e.g. deleting of expired Trusted (Device) Certificates)
* API discovery
* Rapid prototyping (setting up demos, simulating device data)

I look forward to reading your feedback.

**Comment by switschel** [comment]
One addition: Don't hesitate to share some examples you like most ;-)

**Reply by reubenmiller** [reply]
Extra points for the best one-liner!

**Comment by haraldmeyer** [comment]
* Fetch MEA to debug Analytics Builder model
* Create MEA to simulate device for Analytics Builder model 
* Create microservice application for local development

**Comment by reubenmiller** [comment]
Category: Troubleshooting IoT Solution
Use-case: Find devices which don't have an external identity (due to a programming error)

```sh
c8y devices list -p 2000 \
| c8y identity list --outputTemplate "if std.length(output) == 0 then input.value else ''"

# Though obviously you probably want to narrow down your search on the c8y devices list first ;)
c8y devices list --query "has(myFragment) and name eq 'test*'" \
| c8y identity list --outputTemplate "if std.length(output) == 0 then input.value else ''"
```

**Comment by mvoigt-sag** [comment]
- bulk delete filtered devices
- get credentials for bootstrap user
- write extension for custom microservice API (https://github.com/SoftwareAG/c8y-oee)
- write extension for CI/CD build pipeline (https://github.com/SoftwareAG/c8y-bpl)
- replace usage of PostMan almost altogether

**Comment by k-butz** [comment]
- Debugging Platform Data
- Scripting Device Simulators for product demos
- I have some nightly go-c8y-cli jobs running for e.g. custom retention rules or data quality checks
- Browsing Audit Logs (very useful in combination with grepping for keywords)
- Notification 2.0 subscriptions
- Completely replaced Postman for me (good to have `c8y api` as last resort 😄 ) 
- I really like the new Output Templates for adhoc-reporting:
```
-> c8y devices list --type demo \
    | c8y measurements list \
        --withTotalPages \
        --pageSize 1 \
        --dateFrom -1d \
        --outputTemplate "input.value + output.statistics" \
        --select deviceId:id, deviceName:name, otherDeviceProperty:propertyA, ctMeasurements24h:totalPages \
        -o csvheader
deviceId,deviceName,otherDeviceProperty,ctMeasurements24h
74151200,dev_0000001,valueA,691
40150201,dev_0000002,valueA,1984
52151201,dev_0000003,valueA,1854
41152200,dev_0000004,valueA,785
89152201,dev_0000005,valueA,0
```

**Comment by scfx** [comment]
- Create device simulators (main use case).
- Understand platform data.
- Upload microservices.
- Perform bulk deletion, update, and creation of components.
- Set environment variables for UI development.

**Comment by thomaswinkler** [comment]
General use for Github automation to replace custom actions or shell scripting.

**Comment by MarkJMReynolds-SAG** [comment]
Scripts to set up our environments for testing, including (but not limited to)
- Creating new subtenants
- Subscribing to applications
- Adding new roles and setting permissions 
- Adding new users and granting them access to roles and applications
- Installing and updating licenses
- Adjusting tenant settings
- Verifying if all the above is set up as expected 😉
